distributed/shuffle/_shuffle.py

```diff
@@ -139,58 +140,112 @@ def rearrange_by_column_p2p(
     )
 
 
-class P2PShuffleLayer(SimpleShuffleLayer):
+# TODO remove quotes (requires Python >=3.9)
+_T_Key: TypeAlias = "tuple[str, int] | str"
+_T_LowLevelGraph: TypeAlias = "dict[_T_Key, tuple]"
+
+
+class P2PShuffleLayer(Layer):
     def __init__(
         self,
         name: str,
         column: str,
         npartitions: int,
         npartitions_input: int,
-        ignore_index: bool,
         name_input: str,
-        meta_input: pd.DataFrame,
-        parts_out: list | None = None,
+        parts_out: Iterable | None = None,
         annotations: dict | None = None,
     ):
         check_minimal_arrow_version()
         annotations = annotations or {}
         annotations.update({"shuffle": lambda key: key[1]})
-        super().__init__(
-            name,
-            column,
-            npartitions,
-            npartitions_input,
-            ignore_index,
-            name_input,
-            meta_input,
-            parts_out,
-            annotations=annotations,
-        )
-
-    def get_split_keys(self) -> list:
-        # TODO: This is doing some funky stuff to set priorities but we don't need this
-        return []
+        self.name = name
+        self.column = column
+        self.npartitions = npartitions
+        self.name_input = name_input
+        if parts_out:
+            self.parts_out = set(parts_out)
+        else:
+            self.parts_out = set(range(self.npartitions))
+        self.npartitions_input = npartitions_input
+        super().__init__(annotations=annotations)
 
     def __repr__(self) -> str:
         return (
             f"{type(self).__name__}<name='{self.name}', npartitions={self.npartitions}>"
         )
 
-    def _cull(self, parts_out: list) -> P2PShuffleLayer:
+    def get_output_keys(self) -> set[_T_Key]:
+        return {(self.name, part) for part in self.parts_out}
+
+    def is_materialized(self) -> bool:
+        return hasattr(self, "_cached_dict")
+
+    @property
+    def _dict(self) -> _T_LowLevelGraph:
+        """Materialize full dict representation"""
+        self._cached_dict: _T_LowLevelGraph
+        dsk: _T_LowLevelGraph
+        if hasattr(self, "_cached_dict"):
+            return self._cached_dict
+        else:
+            dsk = self._construct_graph()
+            self._cached_dict = dsk
+        return self._cached_dict
+
+    def __getitem__(self, key: _T_Key) -> tuple:
+        return self._dict[key]
+
+    def __iter__(self) -> Iterator[_T_Key]:
+        return iter(self._dict)
+
+    def __len__(self) -> int:
+        return len(self._dict)
```

Comment:
This subclassing caused another problem so I just copied the relevant code. IMHO most of these methods should be included in some kind of baseclass and not necessarily in the `SimpleShuffleLayer`. I see this being repeated often but fixing this right now is out of scope.